diffrax/_solver/quicsort.py

```diff
@@ -0,0 +1,250 @@
+    a_div_h: PyTree[ArrayLike]  # gamma.shape + (1, *taylor)
+    dtype: jnp.dtype = eqx.field(static=True)
+
+    def __init__(self, beta_lr1, a_lr1, b_lr1, a_third, a_div_h):
+        self.beta_lr1 = beta_lr1
+        self.a_lr1 = a_lr1
+        self.b_lr1 = b_lr1
+        self.a_third = a_third
+        self.a_div_h = a_div_h
+        all_leaves = jtu.tree_leaves(
+            [self.beta_lr1, self.a_lr1, self.b_lr1, self.a_third, self.a_div_h]
+        )
+        self.dtype = jnp.result_type(*all_leaves)
+
+
+class QUICSORT(AbstractFosterLangevinSRK[_QUICSORTCoeffs, None]):
+    r"""The QUadrature Inspired and Contractive Shifted ODE with Runge-Kutta Three
+    method by James Foster and Daire O'Kane. This is a third order solver for the
+    Underdamped Langevin Diffusion, the terms for which can be created using
+    [`diffrax.make_underdamped_langevin_term`][]. Uses two evaluations of the vector
+    field per step.
+
+    ??? cite "Reference"
+
+        This is a variant of the SORT method from Definition 1.2 of
+
+        ```bibtex
+        @misc{foster2021shiftedode,
+            title={The shifted ODE method for underdamped Langevin MCMC},
+            author={James Foster and Terry Lyons and Harald Oberhauser},
+            year={2021},
+            eprint={2101.03446},
+            archivePrefix={arXiv},
+            primaryClass={math.NA},
+            url={https://arxiv.org/abs/2101.03446},
+        }
+        ```
+
+        with the modifications inspired by the UBU method from
+
+        ```bibtex
+        @misc{chada2024unbiased,
+            title={Unbiased Kinetic Langevin Monte Carlo with Inexact Gradients},
+            author={Neil K. Chada and Benedict Leimkuhler and Daniel Paulin
+                and Peter A. Whalley},
+            year={2024},
+            eprint={2311.05025},
+            archivePrefix={arXiv},
+            primaryClass={stat.CO},
+            url={https://arxiv.org/abs/2311.05025},
+        }
+        ```
+    """
+
+    interpolation_cls = LocalLinearInterpolation
+    minimal_levy_area = AbstractSpaceTimeTimeLevyArea
+    taylor_threshold: RealScalarLike = eqx.field(static=True)
+
+    def __init__(self, taylor_threshold: RealScalarLike = 0.1):
+        r"""**Arguments:**
+
+        - `taylor_threshold`: If the product `h*gamma` is less than this, then
+        the Taylor expansion will be used to compute the coefficients.
+        Otherwise they will be computed directly. When using float32, the
+        empirically optimal value is 0.1, and for float64 about 0.01.
+        """
+        self.taylor_threshold = taylor_threshold
+
+    def order(self, terms):
+        return 3
+
+    def strong_order(self, terms):
+        return 3.0
+
+    def _directly_compute_coeffs_leaf(
+        self, h: RealScalarLike, c: UnderdampedLangevinLeaf
+    ) -> _QUICSORTCoeffs:
+        del self
+        # compute the coefficients directly (as opposed to via Taylor expansion)
+        original_shape = c.shape
```

Comment:
`jnp.shape(c)`? I don't know if normal Python values can infiltrate this far, but just in case.